services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/validation/ApplicationValidator.kt

```diff
@@ -2,33 +2,87 @@ package fi.hel.haitaton.hanke.validation
+import mu.KotlinLogging
+
+private val logger = KotlinLogging.logger {}
 
 class ApplicationValidator : ConstraintValidator<ValidApplication, Application> {
-    override fun isValid(application: Application?, context: ConstraintValidatorContext?): Boolean {
-
-        val applicationData =
-            application?.applicationData
-                ?: throw InvalidApplicationDataException("Application data missing.")
-
-        return when (applicationData) {
-            is CableReportApplicationData ->
-                listOfNotNull(
-                        applicationData.customerWithContacts.customer,
-                        applicationData.contractorWithContacts.customer,
-                        applicationData.representativeWithContacts?.customer,
-                        applicationData.propertyDeveloperWithContacts?.customer,
-                    )
-                    .all {
-                        when {
-                            it.registryKey == null -> true
-                            it.registryKey.isValidBusinessId() -> true
-                            else -> throw InvalidApplicationDataException("Y-tunnus invalid.")
-                        }
-                    }
+    override fun isValid(application: Application, context: ConstraintValidatorContext?): Boolean {
+
+        return when (application.applicationData) {
+            is CableReportApplicationData -> validOrThrow(application.applicationData)
+        }
+    }
+
+    private fun validOrThrow(cableReportData: CableReportApplicationData): Boolean {
+        if (!isValidCableReport(cableReportData)) {
+            throw InvalidApplicationDataException("Received cable report contains invalid data.")
+        }
+        return true
+    }
+
+    private fun isValidCableReport(cableReportData: CableReportApplicationData): Boolean =
+        with(cableReportData) {
+            listOf(
+                    validStartAndEnd(startTime, endTime),
+                    validCustomerWithContactsNotNull(customerWithContacts),
+                    validCustomerWithContactsNullable(representativeWithContacts),
+                    validCustomerWithContactsNullable(contractorWithContacts),
+                    validCustomerWithContactsNullable(propertyDeveloperWithContacts)
+                )
+                .all { it }
+        }
+
+    /**
+     * Checks if the start is before or equal to the end.
+     *
+     * @param start the start time (nullable)
+     * @param end the end time (nullable)
+     * @return true if start is before or equal to end, false if either is null or if start is after
+     * end.
+     */
+    private fun validStartAndEnd(start: ZonedDateTime?, end: ZonedDateTime?): Boolean {
+        logger.info { "Application startDate: '$start', endDate: '$end'." }
+
+        if (start == null && end == null) {
+            return true
+        }
+
+        if (start == null || end == null) {
+            return false
+        }
```

Comment:
Why?

Reply:
The problem with this is that the user can enter a date for one of the fields and then use the previous- or save and suspend buttons to save the incomplete page. Validating for the missing other value will cause the save to fail. And it's not obvious to the user why this happens.

For proper full validation, we should have one set of validation rules for saving drafts that allow for mandatory values to be missing and one set for sending the application to be processed, which would also check that all mandatory values are entered.

Reply:
```
    private fun validStartAndEnd(start: ZonedDateTime?, end: ZonedDateTime?): Boolean {
        logger.info { "Application startDate: '$start', endDate: '$end'." }

        if (start == null && end == null) {
            return true
        }

        if (start == null || end == null) {
            return false
        }

        return !start.isAfter(end)
    }
```
Both must be null or both must be not null. Do we want to allow only one of them to contain a value?

Reply:
So, what are you suggesting here? Removal of 

```
        if (start == null || end == null) {
            return false
        }
```
or something else?

Reply:
One check with
```kotlin
        if (start == null || end == null) {
            return true
        }
```

Reply:
Done.